zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/LearnerHandler.java

```diff
@@ -615,21 +632,30 @@ public void run() {
             syncThrottler.endSync();
             syncThrottler = null;
 
-            // now that the ack has been processed expect the syncLimit
-            sock.setSoTimeout(learnerMaster.syncTimeout());
-
             /*
              * Wait until learnerMaster starts up
              */
             learnerMaster.waitForStartup();
 
+            int queueSize = queuedPackets.size();
+            while (upToDatePacketsLimit > 0 && queueSize > upToDatePacketsLimit) {
+                LOG.info("Waiting for learner queue to be drained before " +
+                        "sending UPTODATE to {}, current queue size: {}",
+                        sid, queueSize);
+                Thread.sleep(100);
```

Comment:
I am wondering why we need to wait for the learner to drain its queued in flight commits - we could just, as usual, send a `UPTODATE` and then on learner side make sure zookeeper server does not start serving traffic until the queued commits are applied to data tree (which, seems to be what this patch is doing). Do we need wait here?

Reply:
Waiting the queue being drained before sent UPTODATE can help avoid leader timedout when wait on learner's UPDATETO ack due to too many txns queued on leader between NEWLEADER and UPTODATE and need to be replayed on learner.

Reply:
make sense. 

is it possible this loop never finishes? It seems that the `queuedPackets` will be drained but no packets will be further queued in the same time, when we are inside this loop, so it should terminate, but just to double check.